### PR TITLE
Mono 6.12 for DSM 7.0

### PR DESCRIPTION
**Pull request by DigitalBox98**
_Motivation:_  Update mono to 6.12 for DSM 7.0
_Linked issues:_  #4322

### Checklist
- [X] Build rule `all-supported` completed successfully
- [ ] Package upgrade completed successfully
- [ ] New installation of package completed successfully

### CODE REVIEW

**Comment by ymartin59** [comment]
You may exclude arch 88f6281

**Comment by MarkCiliaVincenti** [comment]
Is this ready to get merged to master and released?

File: cross/mono/PLIST [hunk 1]
```diff
@@ -1,94 +1,73 @@
...
+bin:bin/makecert
+bin:bin/mcs
+bin:bin/mdassembler
+bin:bin/mdoc-assemble
+bin:bin/mdoc-export-html
+bin:bin/mdoc-export-msxdoc
+bin:bin/mdoc-update
+bin:bin/mdoc-validate
+bin:bin/mdvalidater
+bin:bin/mkbundle
```
Comment (hgy59) [inline cross/mono/PLIST:17]: please keep files as `rsc` as it is not possible to _strip_  those (and other) files.

File: cross/mono/PLIST [hunk 2]
```diff
@@ -1,94 +1,73 @@
...
+bin:bin/vbc
+bin:bin/wsdl
+bin:bin/wsdl2
+lib:lib/libMonoPosixHelper.so
+lib:lib/libMonoSupportW.so
 lib:lib/libikvm-native.so
-lnk:lib/libmono-2.0.so
-lnk:lib/libmono-2.0.so.1
+lib:lib/libmono-2.0.so
+lib:lib/libmono-2.0.so.1
```
Comment (hgy59) [inline cross/mono/PLIST:49]: please keep `lnk` here (and others in this PLIST) as those are links!

File: spk/mono/Makefile [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
 SPK_ICON = src/mono.png
 
 DEPENDS = cross/$(SPK_NAME)
 
-MAINTAINER = hg59
+MAINTAINER = digitalbox98
 DESCRIPTION = Cross platform, open source .NET development framework
 DISPLAY_NAME = Mono
 STARTABLE = no
-CHANGELOG = "1. Update to version 5.20.1.34<br/>2. Fix workaround for ARMv5 (88f628x archs)"
```
Comment (hgy59) [inline spk/mono/Makefile:12]: please update the CHANGELOG and to not remove it

**Review by hgy59 — CHANGES_REQUESTED** [review]

**Comment by hgy59** [comment]
> Is this ready to get merged to master and released?

No it is not ready yet.

@MarkCiliaVincenti can you tell us, why you would like to have Mono on DSM7?

As I commented in #4524, we do not want to create Mono packages for DSM7, as all packages depending on Mono are outdated and migrated to dotnet (except AFAICR sonarr, but this will only available when migrated to dotnet).

**Comment by MarkCiliaVincenti** [comment]
The issue is indeed Sonarr and it will take a long while before it is
migrated away from the old netframework.

**Comment by picrap** [comment]
> 
> 
> > Is this ready to get merged to master and released?
> 
> No it is not ready yet.
> 
> @MarkCiliaVincenti can you tell us, why you would like to have Mono on DSM7?
> 
> As I commented in #4524, we do not want to create Mono packages for DSM7, as all packages depending on Mono are outdated and migrated to dotnet (except AFAICR sonarr, but this will only available when migrated to dotnet).

We are also still using Mono on DSM7 (not for long I hope), so we also are interested in this.
You mentionned that dotnet is available on DSM7, but where?

**Comment by hgy59** [comment]
> We are also still using Mono on DSM7 (not for long I hope), so we also are interested in this.
> You mentionned that dotnet is available on DSM7, but where?

In this repository dotnet is used as build tool for jellyfin and some *arr-packages.
There is no dotnet runtime or dotnet sdk package for standalone installation available (yet).

I have created such dotnet package (runtime and sdk) some time ago, but never pushed it to this repository.
With DSM 6.1 there where warnings about missing version information for libstdc++, but this was before such a cross package was added to spksrc.

@picrap are you interested in dotnet runtime or dotnet SDK or both? 
Do you want dotnet 3.1 too, or is dotnet 5 all you need?

**Comment by hgy59** [comment]
@MarkCiliaVincenti as for current Sonarr package I think it would be enough to publish the current Mono package (5.20.1.34-17) for DSM7.

**Comment by MarkCiliaVincenti** [comment]
Not sure what you mean.

**Comment by picrap** [comment]
@hgy59 we plan to move to net5 (or later), once we have a successful build for it, but as of today we’re still using mono (that  currently plan to deploy in our package if no Mono package is available for DSM7).

**Comment by hgy59** [comment]
> 
> Not sure what you mean.

The mono packages currently available on https://synocommunity.com/packages are built for DSM6 and are not compatible with DSM7 (despite you can see the package in DSM 7 package center on your diskstation, when you added synocommunity package repository).

The packages for DSM7 must specify a DSM7 revision number (i.e. >= 40000) and must not require root permissions (as our DSM6 packages have by default).

By building the current mono package to target DSM7 we will get DSM7 compatible packages.

**Comment by hgy59** [comment]
I have created mono for DSM7. You can download the packages here: https://github.com/hgy59/spksrc/releases/tag/mono-5.20.1.34-17dsm7

But non of the depenent packages is updated for DSM7 yet (jackett, lidarr, sonarr). 
jackett is updated for DSM7, but only jackett created with dotnet, that does not depend on mono anymore.

**Comment by picrap** [comment]
Thanks for the package! Do you plan to publish it on synocommunity repo?

**Comment by hgy59** [comment]
> Thanks for the package! Do you plan to publish it on synocommunity repo?

Yes I will publish those, when there is no negative feedback...
But first we need some experiance with the dependent packages on DSM7.

**Comment by picrap** [comment]
A simple note regarding the package, `mono` is now located in `/var/packages/mono/target/bin/mono` instead of `/var/packages/Mono/target/usr/local/bin/mono`.

**Comment by picrap** [comment]
So far so good. My package runs with the provided `mono` (AMD64) without problem.

**Comment by publicarray** [comment]
@hgy59 I've recently pushed sonarr and lidarr for DSM7 (dotnet versions only). It did not require any code changes. The first start might take extra long to start since it has to run the updater. (I've tested the downgrade/upgrade path with no issues)

**Comment by hgy59** [comment]
> Thanks for the package! Do you plan to publish it on synocommunity repo?

DSM7 versions of Mono 5.20.1.34-17 are now published. Within the regular delay of up to 48 hours these will be visible in DSM package manager.

**Comment by publicarray** [comment]
@hgy59 I can confirm Sonarr works with your mono version from your fork. But when I compiled mono from master I got the following:

```
su sc-nzbdrone -l -s /bin/bash
bash-4.4 $ env PATH=/var/packages/mono/target/bin:${PATH} HOME=/var/packages/nzbdrone/var LD_LIBRARY_PATH=/var/packages/nzbdrone/target/lib /var/packages/mono/target/bin/mono /var/packages/nzbdrone/target/share/Sonarr/Sonarr.exe
Corlib not in sync with this runtime: The runtime did not find the mscorlib.dll it expected. Expected interface version B886E13F-6276-4BE5-85F8-C6BF5EDBD200 but found 1A5E0066-58DC-428A-B21C-0AD6CDAE2789. Check that your runtime and class libraries are matching.
Loaded from: /volume1/@appstore/mono/lib/mono/4.5/mscorlib.dll
Download a newer corlib or a newer runtime at http://www.mono-project.com/download.
```

To fix it I needed to clean `native/mono`  can we make a note of this?

**Comment by hgy59** [comment]
> To fix it I needed to clean `native/mono` can we make a note of this?

For publishing I always use the same dedicated clone and I always use `make native-clean` before building packages that depend on natives.
Sometime it's even better to use `make dist-clean` to clean everything and build from scratch.
_If a local build fails, but github action succeeds, you better build from scratch._

Another point is, that native packages with equivalent cross packages must have the same version. This is obvious and must be done on package updates, and a version check could be added to the cross makefiles.

**Comment by publicarray** [comment]
@hgy59 Thanks, Yea I guess I was more surprised that the error is only visible at runtime. And since I've never touched mono I did not check that mono had a native component.

File: cross/mono/PLIST (hunk 1 above)
Comment (DigitalBox98) [reply cross/mono/PLIST]: Ok I have fixed the bin as rsc

File: cross/mono/PLIST (hunk 2 above)
Comment (DigitalBox98) [reply cross/mono/PLIST]: indeed, fixed : )

File: spk/mono/Makefile (hunk 3 above)
Comment (DigitalBox98) [reply spk/mono/Makefile:12]: The CHANGELOG now indicates the update for the new version

**Comment by DigitalBox98** [comment]
Ok I have excluded the 88f6281 arch and submitted the various fixes suggested. So far so good, all builds are green : )

**Comment by MarkCiliaVincenti** [comment]
Very stupid question, but will I be able to update simply through the Synology Package Center?

**Comment by Stanzilla** [comment]
Any chance this could be shipped soon? The problems with Sonarr are creeping up again.

**Comment by mreid-tt** [comment]
@publicarray would having a newer Mono version as this PR aims to implement resolve the issues with TLS as reported in #5051 that you were attempting to fix in #5070?

**Comment by hgy59** [comment]
superseded by #5677 that fixes all remaining issues.

**Pull request closed without merging**